### ISSUE TITLE
Help add quizzes!

### ISSUE THREAD

**Issue by kentcdodds** [kentcdodds/es6-workshop#20]
I added a quizzes directory and intend on adding a few quiz functions for every concept to be used before/while teaching the concept. If anyone wants to help make those, please do! I recorded [this video](https://youtu.be/LlNGtPFK9OM) explaining how it works and stuff!

**Comment by alayek** [comment]
Hey @kentcdodds, I am interested in contributing. I have learned a lot from your courses and tutorials, so I would like to help anyway I can 😄 

Before I raise a PR, which is the easy part, I would like to know the following and tune the quizzes accordingly:
- How long each quiz should take an average JS developer to solve?
- What can I assume about the background of the workshop students? Background, as in, prior experience.
- What is the focus of the workshop? I get that ES6 is the main theme, but is the focus "a simple intro to ES6"? Or is it "See how easy ES6 makes your lives easier by reducing boilerplate!"?

I have following ideas for quizzes. I would focus on problems those leverage ES6 syntax to appear short and sweet ✨  ✨ 

- [ ] Fat arrow syntax 🏹 
    - [ ] The syntax is mostly straightforward. Only point to note, is that it binds `this` from its parent scope automagically.
          [This](https://repl.it/FCls/0) should illustrate that. Notice the call to `this.unit` inside the `forEach` loop. In ES5, you would have to specifically bind that, or pass that as an argument to `forEach()`.
    - [ ] Writing currying functions are easier with arrows. I always forget the `return` in one of the nested functions. 🤣 

- [ ] Array Destructuring and new Array methods 🍌 
    - [ ] Write a function `add()` that can take _arbitrary_ numbers of integers as arguments, and return their sum.
           With ES5, there would be some ugly calls to `slice()`.
           With ES6, `const add = (...args) => args.reduce((acc, item) => acc + item)`
    - [ ] Write a function `range(a, b)` to return an array of pattern `[a, a + 1, .., b - 1, b]`. Assume a and b > 0
           With ES5, one would probably have to use a for-loop, or some loadash / underscore utility.
           With ES6, `const range = (a, b) => [...Array(b + 1).keys()].slice(a, b + 1)`
    - [ ] Write a function `randgen(n)` to return an array of size `n` with random integers <= 1000
            With ES6 `Array.from()`, 
            it's `const randgen(n) = Array.from({ length : n }, () => Math.floor(Math.random() * 1000.0))`
    - [ ] Write a function `union(a, b)` to unite array a and b, with no repetition. As in, if both a and b has an element, it should appear only once in the resultant array.
           With ES6 `Array.includes()`, it would be 
          ```js
            const union = (a, b) => a.filter(x => !b.includes(x)).concat(b)
          ```
           Another possible solution that may give slightly different answer is:
           ```js
           const union = (a, b) => a.concat(b.filter(x => !a.includes(x)))
            ```

- [ ] Template Literals 🥇 
    - [ ] Quiz should make it evident that whenever one is trying to print a variable as part of string, it's easier to use template literal. An example like [this](https://repl.it/FCmV/0) would be interesting.

I left out generators, object spread, default parameters etc. I will get to them soon.

But please let me know what do you think? Am I going in the right direction, or is this completely orthogonal to your goals?

One thing I would like to do in the quizzes - if I introduce a concept in a quiz, use the next quizzes to repeat the concept. It helps a beginner assimilate a new concept. Like, if I have introduced a concept of arrow functions, all next quizzes should use arrow function as much as possible.

**Comment by alayek** [comment]
Hi @kentcdodds , I have learned so much about ES6 and Webpack from your guides and tutorials. I would like to contribute and help anyway I can.

Where should I start?

---

I already checked the [list of features](https://github.com/lukehoban/es6features). Do you feel creating an epic would help?

**Comment by kentcdodds** [comment]
Hey @alayek!
Thanks for all of the contributions! This is super!

Before you get too far down on these contributions, let me explain a little bit about what I'm looking for. I'm going to be using these quizzes as a way to teach. So during the quizzes portion of the workshop, everyone's attention will be on me. They wont pull up the code examples themselves. And I'll show them a quiz and ask them what they thing it will return based on the code in there (like the `let`/`const` examples) or I'll tell them that we can refactor one thing to another and ask them to think what the refactored state will look like. This will challenge their intuition and then I'll demonstrate with the quiz function how things work which will confirm or correct their intuition.

So it's important that the comments/examples for each quiz facilitate this type of interaction. They wont be changing the code themselves. There's no need for tests or anything. It will be an interactive thing where I'm driving the keyboard and they're watching. Then when I'm through, they'll go to the tests in `exercises` to do things on their own. So, the best way to think of quiz ideas is to look at the tests and build quizzes based on what they'll be required to do in the tests.

I hope that's helpful. I'll give your PRs a look when I get the chance (soon hopefully). Thank you!

**Comment by kentcdodds** [comment]
Oh, and don't forget to [add yourself to the contributors table](https://github.com/kentcdodds/es6-workshop/blob/master/CONTRIBUTING.md#add-yourself-as-a-contributor)! Thanks!

**Comment by alayek** [comment]
Thanks @kentcdodds ! I will raise a separate PR for adding myself to contributors table 😸

**Comment by kentcdodds** [comment]
Hey @alayek,

Thanks so much for all of your contributions so far! I just pushed [a commit](https://github.com/kentcdodds/es6-workshop/commit/b6ccf9ebacf7655c8c9e147620db5c903924039b) that made a few changes to your quizzes and added a few thing that I think you'll find interesting. A few specific things to note:

1. Put solutions at the bottom
2. I added eslint checking to the quiz files. If there's a reason you need to disable something (to demonstrate the feature or something) then take a look at the bottom of `07_parameters.js` to see how to do that. Probably the biggest two things that I changed about your examples was removing semicolons and using function declarations rather than arrow functions for the helpers.
3. Don't use `.test` in the filename.
4. Have each quiz return something that can be `console.log`ed
5. In the `console.log` at the top, put each quiz on its own line with a trailing comma (see the other quizzes for an example of what I mean)

Also, make sure that things actually run. One of the quizzes had a syntax error I didn't spot when I merged it 😅. Just run `npm run quiz` and have things run as you work.

Thanks again so much for all your help! You've been super! 🎉 

Oh, and don't forget to add yourself to the contributors table! You deserve it!
